README.md
# Minecraft Server Status

[![Latest Stable Version](https://poser.pugx.org/lukasss93/minecraft-server-status/v/stable)](https://packagist.org/packages/lukasss93/minecraft-server-status)
[![Total Downloads](https://poser.pugx.org/lukasss93/minecraft-server-status/downloads)](https://packagist.org/packages/lukasss93/minecraft-server-status)
[![License](https://poser.pugx.org/lukasss93/minecraft-server-status/license)](https://packagist.org/packages/lukasss93/minecraft-server-status)
![PHP](https://img.shields.io/badge/php-%3E%3D5.6-green.svg)

> This library can be used to check Minecraft Servers Status for some basic information.

**⚠ Please do not create issues when you are unable to retrieve information from a server, unless you can prove that there is a bug within the library.**

## Differences between Ping and Query
There are two methods of retrieving information about a Minecraft server.

* **Ping**

    Ping protocol was added in Minecraft 1.7 and is used to query the server for minimal amount of information (hostname, motd, icon, and a sample of players). 
    This is easier to use and doesn't require extra setup on server side. 
    It uses TCP protocol on the same port as you would connect to your server an optional parameter `IsOld17` which can be used to query servers on version 1.6 or older.
    *N.B.: this method doesn't work with Minecraft: Bedrock Edition*

* **Query**

    This method uses GameSpy4 protocol, and requires enabling `query` listener in your `server.properties` like this:

    >*enable-query=true*
    
    >*query.port=25565*

    Query allows to request a full list of servers' plugins and players, however this method is more prone to breaking, so if you don't need all this information, stick to the ping method as it's more reliable.

Requirements
---------
* PHP >= 5.6
* Json Extension
* Iconv Extension

Installation
---------
You can install this library with composer:

`composer require lukasss93/minecraft-server-status`


Using
---------
### MCPing
#### Using
```php
<?php
	use MCServerStatus\MCPing;
	print_r( MCPing::check('hostname or IP') );
?>
```

If you want to get `ping` info from a server that uses a version older than Minecraft 1.7,
then add true parameter after `$timeout`.

Please note that this library does resolve **SRV** records too.

#### Input
The `check()` method has 4 optional parameters:

\# | Parameter | Type | Default |Description
---|-----------|------|---------|-----------
1 | host | string | 127.0.0.1 |Server Hostname or IP address
2 | port | int| 25565 | Server port
3 | timeout | int | 2 | Timeout (in seconds)
4 | isOld17 | bool | false | Boolean value to find informations on servers that uses a version older than Minecraft 1.7

#### Output
The `check()` method return an object with the following properties:

Key|Type|Description
---|----|------------
online|bool|Returns `true` if the server is online else `false`
error|string|Returns any error message
hostname|string|Returns the server hostname or IP address 
address|string|Returns server IP address
port|int|Returns the server port
ping|int|Returns server ping
version|string|Returns the server version
protocol|int|Returns the server protocol
players|int|Returns the number of online players
max_players|int|Returns the maximum number of players that can enter the server
sample_player_list|array|Returns a partial list of online players
motd|string|Returns server description
favicon|string|Returns an image in Base64 string
mods|array|Returns a list of installed mods on the server

_You can use the following methods after_ `check()` _method:_

Method|Description
------|-----------
toArray()|Return the object properties as an array
getMotdToText()|Get the motd without the format codes
getMotdToHtml()|Get the motd as HTML


----

### MCQuery
#### Using
```php
<?php
	use MCServerStatus\MCQuery;
	print_r( MCQuery::check('hostname or IP') );
?>
```

#### Input
The `check()` method has 4 optional parameters:

\# | Parameter | Type | Default |Description
---|-----------|------|---------|-----------
1 | host | string | 127.0.0.1 | Server Hostname or IP address
2 | port | int| 25565 | Server query port
3 | timeout | int | 2 | Timeout (in seconds)
4 | resolveSRV | bool | true | Resolve SRV record

#### Output
The `check()` method return an array with the following properties:

Key|Type|Description
---|----|------------
online|bool|Returns `true` if the server is online else `false`
error|string|Returns any error message
hostname|string|Returns the server hostname or IP address
address|string|Returns server IP address
port|int|Returns the server port
version|string|Returns the server version
software|string|Returns the server software
game_type|string|Returns the server software type
game_name|string|Return the server software name
players|int|Returns the number of online players
max_players|int|Returns the maximum number of players that can enter the server
player_list|array|Returns a list of online players
motd|string|Returns server description
map|string|Returns the server map name
plugins|array|Returns a list of installed plugins on the server

_You can use the following methods after_ `check()` _method:_

Method|Description
------|-----------
toArray()|Return the object properties as an array
getMotdToText()|Get the motd without the format codes
getMotdToHtml()|Get the motd as HTML

Changelog
---------
All notable changes to this project will be documented [here](https://github.com/Lukasss93/minecraft-server-status/blob/master/CHANGELOG.md).

### Recent changes
## [2.0]
- First release


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [2.0]
- First release


src/MCPing.php
<?php

namespace MCServerStatus;

use Exception;
use MCServerStatus\Exceptions\MCPingException;
use MCServerStatus\Responses\MCPingResponse;

/**
 * PHP class to check the status of a Minecraft server by Ping
 * Class MCPing
 * @package MCServerStatus
 */
class MCPing {
	
	private static $socket;
	private static $timeout;
	private static $response;
	
	private function __construct() {
	}
	
	/**
	 * Check and get the server status
	 * @param string $hostname
	 * @param int    $port
	 * @param int    $timeout
	 * @param bool   $isOld17
	 * @return MCPingResponse
	 */
	public static function check($hostname = '127.0.0.1', $port = 25565, $timeout = 2, $isOld17 = false) {
		
		//initialize response
		self::$response = new MCPingResponse();
		
		try {
			
			//save hostname
			self::$response->hostname = $hostname;
			
			//check port
			if(!is_int($port) || $port < 1024 || $port > 65535) {
				throw new MCPingException('Invalid port');
			}
			self::$response->port = $port;
			
			//check timeout
			if(!is_int($timeout) || $timeout < 0) {
				throw new MCPingException('Invalid timeout');
			}
			self::$timeout = $timeout;
			
			//validate isold17 parameter
			if(!is_bool($isOld17)) {
				throw new MCPingException('Invalid parameter in $isold17');
			}
			
			//validate host
			if(filter_var(self::$response->hostname, FILTER_VALIDATE_IP)) {
				//host is ip => address is host
				self::$response->address = self::$response->hostname;
			}
			else {
				//find domain ip
				$resolvedIP = gethostbyname(self::$response->hostname);
				
				if(filter_var($resolvedIP, FILTER_VALIDATE_IP)) {
					//resolvedIP is a valid IP => address is resolvedIP
					self::$response->address = $resolvedIP;
				}
				else {
					//resolvedIP is not a valid IP then find SRV record
					$dns = @dns_get_record('_minecraft._tcp.' . self::$response->hostname, DNS_SRV);
					/*
					 * added @ because in a few server, php return this error:
					 * 'dns_get_record(): A temporary server error occurred'
					 * the resolution time is almost always 15-25sec.
					 * I don't know how fix the time.
					 */
					
					if(!$dns) {
						throw new MCPingException('dns_get_record(): A temporary server error occurred');
					}
					
					if(is_array($dns) and count($dns) > 0) {
						self::$response->address = gethostbyname($dns[0]['target']);
						self::$response->port = $dns[0]['port'];
					}
				}
			}
			
			//open the socket
			self::$socket = @fsockopen(self::$response->address, self::$response->port, $errno, $errstr, self::$timeout);
			if(!self::$socket) {
				throw new MCPingException("Failed to connect or create a socket: $errstr");
			}
			stream_set_timeout(self::$socket, self::$timeout);
			
			//get server infos
			if(!$isOld17) {
				self::ping();
			}
			else {
				self::pingOld();
			}
			
			//server status (if you are here the server is online obv)
			self::$response->online = true;
		}
		catch(MCPingException $e) {
			self::$response->error = $e->getMessage();
		}
		catch(Exception $e) {
			self::$response->error = $e->getMessage();
		}
		finally {
			//close the socket
			@fclose(self::$socket);
		}
		
		//return the response
		return self::$response;
	}
	
	/**
	 * Ping the server with version >= 1.7
	 * @throws MCPingException
	 */
	private static function ping() {
		
		// for read timeout purposes
		$timestart = microtime(true);
		
		//data to send with socket, see http://wiki.vg/Protocol (Status Ping)
		
		//packet ID = 0 (varint)
		$data = "\x00";
		
		//protocol version (varint)
		$data .= "\x04";
		
		//server (varint len + UTF-8 addr)
		$data .= pack('c', strlen(self::$response->address)) . self::$response->address;
		
		//server port (unsigned short)
		$data .= pack('n', self::$response->port);
		
		//next state: status (varint)
		$data .= "\x01";
		
		//prepend length of packet ID + data
		$data = pack('c', strlen($data)) . $data;
		
		//handshake
		fwrite(self::$socket, $data);
		
		//start ping time
		$startPing = microtime(true);
		
		// status ping
		fwrite(self::$socket, "\x01\x00");
		
		//full packet length
		$length = self::readVarInt();
		if($length < 10) {
			throw new MCPingException('Response length not valid');
		}
		
		//packet type, in server ping it's 0
		fgetc(self::$socket);
		
		//string length
		$length = self::readVarInt();
		$data = "";
		do {
			if(microtime(true) - $timestart > self::$timeout) {
				throw new MCPingException('Server read timed out');
			}
			$remainder = $length - strlen($data);
			
			//get the json string
			$block = fread(self::$socket, $remainder);
			
			//abort if there is no progress
			if(!$block) {
				throw new MCPingException('Server returned too few data');
			}
			$data .= $block;
		}
		while(strlen($data) < $length);
		
		//calculate the ping
		self::$response->ping = round((microtime(true) - $startPing) * 1000);
		
		//no data
		if($data === false) {
			throw new MCPingException('Server didn\'t return any data');
		}
		
		//decode json data
		$data = json_decode($data, true);
		
		//optional json error
		if(json_last_error() !== JSON_ERROR_NONE) {
			throw new MCPingException(json_last_error_msg());
		}
		
		self::$response->version = $data['version']['name'];
		self::$response->protocol = $data['version']['protocol'];
		self::$response->players = $data['players']['online'];
		self::$response->max_players = $data['players']['max'];
		self::$response->sample_player_list = self::createSamplePlayerList(@$data['players']['sample']);
		self::$response->motd = self::createMotd($data['description']);
		self::$response->favicon = isset($data['favicon']) ? $data['favicon'] : null;
		self::$response->mods = isset($data['modinfo']) ? $data['modinfo'] : null;
	}
	
	/**
	 * Ping the server with version < 1.7
	 * @throws MCPingException
	 */
	private static function pingOld() {
		fwrite(self::$socket, "\xFE\x01");
		
		//start ping time
		$startPing = microtime(true);
		
		$data = fread(self::$socket, 512);
		self::$response->ping = round((microtime(true) - $startPing) * 1000);
		
		$length = strlen($data);
		if($length < 4 || $data[0] !== "\xFF") {
			throw new MCPingException('$length < 4 || $data[ 0 ] !== "\xFF"');
		}
		
		//strip packet header (kick message packet and short length)
		$data = substr($data, 3);
		$data = iconv('UTF-16BE', 'UTF-8', $data);
		
		//are we dealing with Minecraft 1.4+ server?
		if($data[1] === "\xA7" && $data[2] === "\x31") {
			$data = explode("\x00", $data);
			self::$response->motd = $data[3];
			self::$response->players = intval($data[4]);
			self::$response->max_players = intval($data[5]);
			self::$response->protocol = intval($data[1]);
			self::$response->version = $data[2];
		}
		else {
			$data = explode("\xA7", $data);
			self::$response->motd = substr($data[0], 0, -1);
			self::$response->players = isset($data[1]) ? intval($data[1]) : 0;
			self::$response->max_players = isset($data[2]) ? intval($data[2]) : 0;
			self::$response->protocol = 0;
			self::$response->version = '1.3';
		}
		
	}
	
	/**
	 * Read int var from socket
	 * @return int
	 * @throws MCPingException
	 */
	private static function readVarInt() {
		$i = 0;
		$j = 0;
		
		while(true) {
			$k = @fgetc(self::$socket);
			if($k === false) {
				return 0;
			}
			$k = ord($k);
			$i |= ($k & 0x7F) << $j++ * 7;
			if($j > 5) {
				throw new MCPingException('VarInt too big');
			}
			if(($k & 0x80) != 128) {
				break;
			}
		}
		
		return $i;
	}
	
	/**
	 * Return sample player list
	 * @param $obj
	 * @return array|null
	 */
	private static function createSamplePlayerList($obj) {
		return (isset($obj) && is_array($obj) && count($obj) > 0) ? $obj : null;
	}
	
	/**
	 * Build the motd
	 * @param $string
	 * @return mixed|string
	 */
	private static function createMotd($string) {
		if(!is_array($string)) {
			return $string;
		}
		else if(isset($string['extra'])) {
			$output = '';
			
			foreach($string['extra'] as $item) {
				if(isset($item['color'])) {
					switch($item['color']) {
						case 'black':
							$output .= '§0';
							break;
						case 'dark_blue':
							$output .= '§1';
							break;
						case 'dark_green':
							$output .= '§2';
							break;
						case 'dark_aqua':
							$output .= '§3';
							break;
						case 'dark_red':
							$output .= '§4';
							break;
						case 'dark_purple':
							$output .= '§5';
							break;
						case 'gold':
							$output .= '§6';
							break;
						case 'gray':
							$output .= '§7';
							break;
						case 'dark_gray':
							$output .= '§8';
							break;
						case 'blue':
							$output .= '§9';
							break;
						case 'green':
							$output .= '§a';
							break;
						case 'aqua':
							$output .= '§b';
							break;
						case 'red':
							$output .= '§c';
							break;
						case 'light_purple':
							$output .= '§d';
							break;
						case 'yellow':
							$output .= '§e';
							break;
						case 'white':
							$output .= '§f';
							break;
					}
				}
				
				if(isset($item['obfuscated'])) {
					$output .= '§k';
				}
				
				if(isset($item['bold'])) {
					$output .= '§l';
				}
				
				if(isset($item['strikethrough'])) {
					$output .= '§m';
				}
				
				if(isset($item['underline'])) {
					$output .= '§n';
				}
				
				if(isset($item['italic'])) {
					$output .= '§o';
				}
				
				if(isset($item['reset'])) {
					$output .= '§r';
				}
				
				$output .= $item['text'];
			}
			
			if(isset($string['text'])) {
				$output .= $string['text'];
			}
			
			return $output;
		}
		else if(isset($string['text'])) {
			return $string['text'];
		}
		else {
			return $string;
		}
	}
}